016-kafka-connect-build.md

```diff
@@ -0,0 +1,127 @@
+
+## Proposal
+
+A new section named `build` would be added to the `KafkaConnect` custom resource.
+This section will allow users to configure:
+* connector plugins
+* build output
+
+The connector plugins section will be a list where one or more connectors can be defined.
+Each connector consists of a name and list of artifacts which should be downloaded.
+The artifacts can be of different types.
+For example:
+* `jar` for a directly download of Java JARs
+* `tgz` or `zip` to _download and unpack_ the artifacts
+* `maven` to download JARs from Maven based on Maven coordinates
+* and possibly others if required
+
+The second part - `output` - will configure how the newly built image will be handled.
+It will support two types:
+* `docker` to push the image to any Docker compatible container repository.
+* `imagestream` to push the image to OpenShift ImageStream (supported only on OpenShift)
+
+Strimzi will not run its own container repository.
+So users of the `docker` type build will need to provide credentials to their own Docker compatible container registry.
+This does not have to run within the same Kubernetes cluster - it can be also SaaS registry such as Docker Hub or Quay.io.
+
+Following is an example of the Kafka Connect CR:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+metadata:
+  name: my-connect-cluster
+  annotations:
+    strimzi.io/use-connector-resources: "true"
+spec:
+  version: 2.6.0
+  replicas: 1
+  bootstrapServers: my-cluster-kafka-bootstrap:9093
+  build:
+    output:
+      type: docker
+      image: docker.io/user/image:tag
+      pushSecret: dockerhub-credentaials
+    plugins:
+      - name: my-connector
+        artifacts:
+          - type: jar
+            url: https://some.url/artifact.jar
+            sha512sum: abc123
+          - type: jar
+            url: https://some.url/artifact2.jar
+            sha512sum: def456
+      - name: my-connector2
+        # ...
+  tls:
+    trustedCertificates:
+      - secretName: my-cluster-cluster-ca-cert
+        certificate: ca.crt
+  config:
+    group.id: connect-cluster
+    offset.storage.topic: connect-cluster-offsets
+    config.storage.topic: connect-cluster-configs
+    status.storage.topic: connect-cluster-status
+```
+
+When the `build` section is specified in the `KafkaConnect` resource, Strimzi will generate `Dockerfile` which will correspond to the configured connectors and their artifacts.
+It will build the container image from this `Dockerfile` in a separate pod and push it to the configured output (Docker registry / ImageStream).
+It will take the digest of this image and configure the Kafka Connect cluster to use this image.
+
+The build will use two different technologies:
+* [Kaniko](https://github.com/GoogleContainerTools/kaniko) will be used on Kubernetes
```

Comment:
just to be clear, it's on user shoulders to run this Kaniko on the same Kubernetes cluster, right? It's not Strimzi going to deploy it.

Reply:
Kaniko is basically just an executable / container image which can build Dockerfiles. Strimzi will create a pod which uses it, pass the Dockerfile to it and it builds the new image and pushes it to the registry. There is nothing to be deployed by the users since it is not some complicated framework which needs to be installed upfront.